pkg/api/server/mock/backend.go

```diff
@@ -0,0 +1,733 @@
+			return systemRecord
+		}
+	}
+
+	return nil
+}
+
+func (backend *Backend) newMockBuild(systemID v1.SystemID, v v1.SystemVersion) *v1.Build {
+	service1Path := tree.Path(fmt.Sprintf("/%s/api", systemID))
+	build := &v1.Build{
+		ID:      v1.BuildID(uuid.NewV4().String()),
+		State:   v1.BuildStatePending,
+		Version: v,
+		Services: map[tree.Path]v1.ServiceBuild{
+			service1Path: {
+				ContainerBuild: v1.ContainerBuild{
+					State: v1.ContainerBuildStatePending,
+				},
+			},
+		},
+	}
+
+	return build
+}
+
+func (backend *Backend) runBuild(build *v1.Build) {
+	// try to simulate reality by making things take a little longer. Sleep for a bit...
+	time.Sleep(2 * time.Second)
+
+	// change state to running
+	build.State = v1.BuildStateRunning
+	now := time.Now()
+	build.StartTimestamp = &now
+
+	// run service builds
+	for sp, s := range build.Services {
+		s.State = v1.ContainerBuildStateRunning
+		s.StartTimestamp = &now
+		s.ContainerBuild.State = v1.ContainerBuildStateRunning
+		s.ContainerBuild.StartTimestamp = &now
+
+		build.Services[sp] = s
+	}
+
+	// sleep
+	fmt.Printf("Mock: Running build %s. Sleeping for 7 seconds\n", build.ID)
+	time.Sleep(7 * time.Second)
+	backend.finishBuild(build)
+
+}
+
+func (backend *Backend) finishBuild(build *v1.Build) {
+	// change state to succeeded
+	now := time.Now()
+
+	// finish service builds
+	for sp, s := range build.Services {
+		s.State = v1.ContainerBuildStateSucceeded
+		s.CompletionTimestamp = &now
+		s.ContainerBuild.CompletionTimestamp = &now
+		s.ContainerBuild.State = v1.ContainerBuildStateSucceeded
+		s.CompletionTimestamp = &now
+		build.Services[sp] = s
+	}
+
+	build.CompletionTimestamp = &now
+	build.State = v1.BuildStateSucceeded
+
+	fmt.Printf("Build %s finished\n", build.ID)
+
+	// update system services...
+	systemRecord := backend.getSystemRecordForBuild(build.ID)
+	services := make(map[tree.Path]v1.Service)
+	for _, build := range systemRecord.builds {
+		for path := range build.Services {
+			services[path] = v1.Service{
+				ID:                 v1.ServiceID(uuid.NewV4().String()),
+				State:              v1.ServiceStateStable,
+				Instances:          []string{uuid.NewV4().String()},
+				Path:               path,
+				AvailableInstances: 1,
+			}
+		}
+		break
+	}
+
+	systemRecord.system.Services = services
+
+}
+
+func (backend *Backend) monitorDeployBuild(deploy *v1.Deploy, buildID v1.BuildID, systemID v1.SystemID) {
+
+	for i := 0; i <= 200; i++ {
+		build, _ := backend.GetBuild(systemID, buildID)
+		newState := getDeployStateByBuild(build.State)
+		if deploy.State != newState {
+			deploy.State = newState
+		}
+
+		if deploy.State == v1.DeployStateSucceeded || deploy.State == v1.DeployStateFailed {
+			return
+		}
+
+		time.Sleep(200 * time.Millisecond)
+	}
+
+	// timed out... fail
+	deploy.State = v1.DeployStateFailed
+
+}
+
+func getDeployStateByBuild(buildState v1.BuildState) v1.DeployState {
+	switch buildState {
```

Comment:
still a bit of a mismatch here:

currently the semantics are:
- there can only ever be one deploy per system at a time that is in accepted or in-progress
- a deploy is first set into pending
- a lock is grabbed, and if there aren't any other deploys in accepted or in-progress, the deploy can move to accepted. if there are, the deploy moves to failed
- the deploy then stays in accepted until its build has succeeded or failed. if the build fails, the deploy moves to failed. if the build succeeds, the deploy moves from accepted to in progress
- the deploy can then succeed or fail

sorry my prior explanations were much more confusing. does this make sense?

this is also subject to change in the future, but i'll be happy to update this to match those semantics as they are introduced